### DISCUSSION TITLE
Fixing frequency measurements with PCNT when using open collector configuration

### DISCUSSION THREAD

**Discussion by danalvarez** [espressif/arduino-esp32#8368] [Show and tell]
Writing this in case it might ever help someone else. Using an ESP32-WROOM-32UE.

I am using the [FreqCountESP](https://github.com/kapraran/FreqCountESP/tree/master) library, which lets you use the PCNT module to count pulses on a pin. I had been testing it with a signal generator connected directly to `IO4`, no problems at all! Very precise and exact, to within 1 Hz which is what I needed. However, I then connected the signal as in the following schematic:

![open_collector_circuit](https://github.com/espressif/arduino-esp32/assets/13237986/8c8f426c-f907-44d3-89d2-ec4e34d71328)

Which is what I was going to ultimately use to sample an RPM signal. To my surprise, frequency measurements became very inexact and noisy. For example, I'd set a 500 Hz test signal on the base of the transistor, and `IO4` would read anything between 500 and 510 Hz. If I disabled the usage of the PCNT module and simply used interruptions to measure the frequency, it would work OK again.

After some reading, I came across some posts mentioning that the glitch filter could help. I configured it by setting `pcnt_set_filter_value(PCNT_UNIT, 1);` and enabled it with `pcnt_filter_enable(PCNT_UNIT);`, and this solved the issue!

Still don't know why, I can't see any glitches with my oscilloscope, signal looks perfectly clean:
![SDS00013](https://github.com/espressif/arduino-esp32/assets/13237986/379f2578-854b-408a-99d7-a2218ea03785)

If anyone has any ideas as to why this solved it, I'd appreciate it!

**Comment by mrengineer7777** [comment]
Conceptually, I see no issues with your circuit.  It does introduce the following changes:

- Signal goes from 5V->3V
- Signal is inverted
- The 1k pullup will cause a rounded rising edge

I suspect your issue is the slower rising edge.  You can see this on your scope measurements: Rise=486ns vs Fall= 82ns.

You can fix this with the following methods:

- Use two transistors in push/pull configuration to get sharp edges
- Use an op-amp instead of a transistor (watch out for 5V->3V level shift)
- Use hysteresis or a slower sampling rate on the software side.  I'm guessing the glitch filter does this.

**Comment by woodlist** [comment]
> pcnt_filter_enable(PCNT_UNIT);

Please note that anybody might not be a programming guru. Where does it need to be added those two operators, in setup, in library?
By the way, this library seems interfere with WiFi library and the MC falls into endless restarts.

**Reply by danalvarez** [reply]
The FreqCountESP library uses the `setupPcnt` function to configure the PCNT module. I've added the filter configurations below. Note that the second argument to `pcnt_set_filter_value(PCNT_UNIT_0, your_window)` would be different depeding on the filter window you want to configure.

```c++
static pcnt_isr_handle_t setupPcnt(uint8_t pin, volatile uint32_t *backupCounter) { 
  pcnt_config_t pcntConfig = {
    .pulse_gpio_num = pin,
    .ctrl_gpio_num = -1,
    .pos_mode = PCNT_CHANNEL_EDGE_ACTION_INCREASE,
    .neg_mode = PCNT_CHANNEL_EDGE_ACTION_HOLD,
    .counter_h_lim = PCNT_HIGH_LIMIT,
    .counter_l_lim = PCNT_LOW_LIMIT,
    .unit = PCNT_UNIT,
    .channel = PCNT_CHANNEL,
  };
  pcnt_unit_config(&pcntConfig);
  pcnt_set_filter_value(PCNT_UNIT_0, 1);                                // ADD THIS TO CONFIGURE THE FILTER
  pcnt_counter_pause(PCNT_UNIT);
  pcnt_counter_clear(PCNT_UNIT);
  pcnt_event_enable(PCNT_UNIT, PCNT_EVT_H_LIM);  // Interrupt on high limit.
  pcnt_isr_handle_t isrHandle;
  pcnt_isr_register(onHLim, (void *)backupCounter, 0, &isrHandle);
  pcnt_intr_enable(PCNT_UNIT);
  pcnt_filter_enable(PCNT_UNIT_0);                                      // ADD THIS TO ENABLE THE FILTER
  pcnt_counter_resume(PCNT_UNIT);
  return isrHandle;
}
```

**Reply by danalvarez** [reply]
As for the library interfering with WiFi, I did not have this experience. I'm assuming the MC is restarting because of core panics. Have you tried decoding the core dump to analyze where it's failing? You can use https://github.com/me-no-dev/EspExceptionDecoder

**Reply by woodlist** [reply]
Thanks for your reply, buddy! I should report that the core panic has not been seen. The serial output behaves like the board is just powered on.

**Reply by danalvarez** [reply]
What's the reset cause? Is it POWER_ON reset? Have you tried another board?

**Reply by woodlist** [reply]
I think the board is ok. Without this lib inclusion everything works as expected. Unfortunately, for the moment I do not have another board. However, I can share the sketch for further in vestigation. Will be available late, this evening.

**Reply by woodlist** [reply]
I have tried to use different libraries for stability increasing. For input debounce eliminating it's commented out the first used "Throttles".  For porting try to Rasperry Pi PicoW has been using EEPROM instead of ESP32 specific "Preferences", which does not do because of Websockets falling. The code allows virtual dashboard connection by Virtuino IOT cross-platform application.

**Reply by danalvarez** [reply]
Given the length of your code it is a little bit difficult to find the culprit. My suggestion would be to make a simpler version of your code with only the core functionality, and check if the resets still occur when you add the FreqCountESP library functionality. Also, check the version of the Arduino core for ESP32 that you are using. I am using 2.0.14.

**Reply by woodlist** [reply]
I have removed everything not related to the library's performance and got the following compilation errors (by the way an hour ago, ESP core on Arduino was updated).
`In file included from c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src/FreqCountESP.h:12,
                 from C:\Users\woodl\AppData\Local\Temp\.arduinoIDE-unsaved2024312-7968-1fdbssb.fc0gi\sketch_apr12a\sketch_apr12a.ino:4:
C:\Users\woodl\AppData\Local\Arduino15\packages\esp32\tools\esp32-arduino-libs\idf-release_v5.1-3662303f31/esp32/include/driver/deprecated/driver/pcnt.h:15:2: warning: #warning "legacy pcnt driver is deprecated, please migrate to use driver/pulse_cnt.h" [-Wcpp]
   15 | #warning "legacy pcnt driver is deprecated, please migrate to use driver/pulse_cnt.h"
      |  ^~~~~~~
In file included from c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.h:12,
                 from c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:1:
C:\Users\woodl\AppData\Local\Arduino15\packages\esp32\tools\esp32-arduino-libs\idf-release_v5.1-3662303f31/esp32/include/driver/deprecated/driver/pcnt.h:15:2: warning: #warning "legacy pcnt driver is deprecated, please migrate to use driver/pulse_cnt.h" [-Wcpp]
   15 | #warning "legacy pcnt driver is deprecated, please migrate to use driver/pulse_cnt.h"
      |  ^~~~~~~
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp: In function 'void onHLim(void*)':
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:25:39: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
   25 |   *(volatile uint32_t *)backupCounter += PCNT_HIGH_LIMIT;
      |                                       ^
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp: At global scope:
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:52:24: warning: ignoring attribute 'section (".iram1.3")' because it conflicts with previous 'section (".iram1.1")' [-Wattributes]
   52 | void IRAM_ATTR onTimer()
      |                        ^
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.h:16:16: note: previous declaration here
   16 | void IRAM_ATTR onTimer();
      |                ^~~~~~~
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:90:23: warning: ignoring attribute 'section (".iram1.4")' because it conflicts with previous 'section (".iram1.0")' [-Wattributes]
   90 | void IRAM_ATTR onRise()
      |                       ^
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.h:15:16: note: previous declaration here
   15 | void IRAM_ATTR onRise();
      |                ^~~~~~
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp: In function 'void onRise()':
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:93:18: warning: '++' expression of 'volatile'-qualified type is deprecated [-Wvolatile]
   93 |   _FreqCountESP::sCount++;
      |   ~~~~~~~~~~~~~~~^~~~~~
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp: In member function 'void _FreqCountESP::_begin(uint8_t, uint8_t)':
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:125:5: error: 'timerAlarmEnable' was not declared in this scope; did you mean 'timerAlarm'?
  125 |     timerAlarmEnable(mTimer);
      |     ^~~~~~~~~~~~~~~~
      |     timerAlarm
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp: In member function 'void _FreqCountESP::begin(uint8_t, uint16_t, uint8_t, uint8_t)':
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:134:22: error: too many arguments to function 'hw_timer_t* timerBegin(uint32_t)'
  134 |   mTimer = timerBegin(hwTimerId, 80, true);
      |            ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from C:\Users\woodl\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.0-rc1\cores\esp32/esp32-hal.h:84,
                 from C:\Users\woodl\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.0-rc1\cores\esp32/Arduino.h:36,
                 from c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.h:4:
C:\Users\woodl\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.0-rc1\cores\esp32/esp32-hal-timer.h:35:14: note: declared here
   35 | hw_timer_t * timerBegin(uint32_t frequency);
      |              ^~~~~~~~~~
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:135:23: error: too many arguments to function 'void timerAttachInterrupt(hw_timer_t*, void (*)())'
  135 |   timerAttachInterrupt(mTimer, &onTimer, true);
      |   ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\woodl\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.0-rc1\cores\esp32/esp32-hal-timer.h:50:6: note: declared here
   50 | void timerAttachInterrupt(hw_timer_t * timer, void (*userFunc)(void));
      |      ^~~~~~~~~~~~~~~~~~~~
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:136:3: error: 'timerAlarmWrite' was not declared in this scope; did you mean 'timerWrite'?
  136 |   timerAlarmWrite(mTimer, timerMs * 1000, true);
      |   ^~~~~~~~~~~~~~~
      |   timerWrite
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp: In member function 'void _FreqCountESP::end()':
c:\Users\woodl\Documents\Arduino\libraries\FreqCountESP\src\FreqCountESP.cpp:172:5: error: 'timerAlarmDisable' was not declared in this scope
  172 |     timerAlarmDisable(mTimer);
      |     ^~~~~~~~~~~~~~~~~`

**Reply by woodlist** [reply]
I have installed overnight build 2.0.15

**Reply by woodlist** [reply]
Seems Espressif guys did an instant revolution. [https://github.com/espressif/arduino-esp32/issues/9169?ysclid=lux27g5kxy803967698](url)

**Comment by RogerDuncan** [comment]
Hi, I stumbled over this old thread because I had a similar problem and want to contribute my findings and **solution**.
I found that all features of the ESP32 CPUs (like PCNT or flank triggered interrupts) that react on flanks of external  signals are problematic **if the flanks are not steep enough**. Even if the signals look perfectly clean on the oscilloscope, it happens that the ESP32 recognizes "too many" events (multiple events on one flank or non existing events on the "wrong" flank (falling flank instead of rising and vice versa). 
I believe that this is because the GPIO pins have no hysteresis and therefore if the signal is somewhere in the middle between low and high even minor noise (even from inside the CPU) can be recognized as a flank. 
**Conclusion**:  When using flank triggered features with ESP32, it is important to make sure that the signal flanks are sufficiently steep. If necessary, a fast Schmitt Trigger device should be added to improve the signal quality.
